### ISSUE TITLE
IndexError due to empty args in OpenSSL.SSL.ZeroReturnError handling in Twisted

### ISSUE THREAD

**Issue by OliverFarren** [twisted/twisted#12181]
**Describe the incorrect behavior you saw**

The application throws an `IndexError` when trying to access the first argument of the OpenSSL.SSL.ZeroReturnError exception's args attribute in the Twisted library. This occurs because ZeroReturnError is raised without any arguments, leading to an empty args tuple, which the code does not currently handle.

**Describe how to cause this behavior**

Set up a TLS connection using Twisted's protocols where the connection might cleanly close, potentially triggering OpenSSL.SSL.ZeroReturnError.
Observe the error handling where Twisted attempts to log or process the error details from exceptionObject.args[0].
This behavior occurs every time a TLS connection is closed in such a manner that it raises ZeroReturnError without any arguments.

```
Traceback (most recent call last):
  File "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py", line 121, in <module>
    errorQueue = exceptionObject.args[0]
IndexError: tuple index out of range
```

**Describe the correct behavior you'd like to see**

The library should handle exceptions that might not have any arguments without crashing. A robust handling might involve checking if `args` is non-empty before attempting to access its contents.

[Link to source code with issue](https://github.com/twisted/twisted/blob/trunk/src/twisted/protocols/tls.py#L120-L122)

**Comment by OliverFarren** [comment]
Not sure if this would be the relevant test for this issue. Not too familiar with the inner-workings of twisted. https://github.com/twisted/twisted/blob/trunk/src/twisted/protocols/test/test_tls.py#L1008

**Comment by glyph** [comment]
> This behavior occurs every time a TLS connection is closed in such a manner that it raises ZeroReturnError without any arguments

Do you have any sense of what the "manner" in question here is?

**Comment by glyph** [comment]
> Not sure if this would be the relevant test for this issue. Not too familiar with the inner-workings of twisted. https://github.com/twisted/twisted/blob/trunk/src/twisted/protocols/test/test_tls.py#L1008

FYI, if you hit <kbd>y</kbd> on any source page on Github, you get a link more like this https://github.com/twisted/twisted/blob/c465c460c68bec3cf481c2815d1039da6796eb03/src/twisted/protocols/test/test_tls.py#L1008 which permanently links to the appropriate version of the file, which means your line-number reference won't break the second someone updates the file 😄

**Comment by OliverFarren** [comment]
> > This behavior occurs every time a TLS connection is closed in such a manner that it raises ZeroReturnError without any arguments
> 
> Do you have any sense of what the "manner" in question here is?

I can provide a detailed stack-trace. Does this help clarify things?

``` json
  "body": {
    "trace_chain": [
      {
        "frames": [
          {
            "code": "why = method()",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/internet/asyncioreactor.py",
            "argspec": [
              "self",
              "selectable",
              "read"
            ],
            "lineno": 138,
            "method": "_readOrWrite",
            "locals": {
              "e": "<class 'IndexError'>",
              "read": true,
              "self": "<class 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'>",
              "method": "<class 'method'>",
              "selectable": "<class 'twisted.internet.tcp.Client'>",
              "why": "<class 'IndexError'>"
            }
          },
          {
            "code": "return self._dataReceived(data)",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/internet/tcp.py",
            "argspec": [
              "self"
            ],
            "lineno": 248,
            "method": "doRead",
            "locals": {
              "self": "<class 'twisted.internet.tcp.Client'>",
              "data": "b'\\x15\\x03\\x03\\x00\\x02\\x01\\x00'"
            }
          },
          {
            "code": "rval = self.protocol.dataReceived(data)",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/internet/tcp.py",
            "argspec": [
              "self",
              "data"
            ],
            "lineno": 253,
            "method": "_dataReceived",
            "locals": {
              "self": "<class 'twisted.internet.tcp.Client'>",
              "data": "b'\\x15\\x03\\x03\\x00\\x02\\x01\\x00'"
            }
          },
          {
            "code": "return self._wrappedProtocol.dataReceived(data)",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/internet/endpoints.py",
            "argspec": [
              "self",
              "data"
            ],
            "lineno": 151,
            "method": "dataReceived",
            "locals": {
              "self": "<class 'twisted.internet.endpoints._WrappingProtocol'>",
              "data": "b'\\x15\\x03\\x03\\x00\\x02\\x01\\x00'"
            }
          },
          {
            "code": "self._checkHandshakeStatus()",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py",
            "argspec": [
              "self",
              "bytes"
            ],
            "lineno": 324,
            "method": "dataReceived",
            "locals": {
              "self": "<class 'twisted.protocols.tls.BufferingTLSTransport'>",
              "bytes": "b'\\x15\\x03\\x03\\x00\\x02\\x01\\x00'"
            }
          },
          {
            "code": "self._tlsShutdownFinished(Failure())",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py",
            "argspec": [
              "self"
            ],
            "lineno": 251,
            "method": "_checkHandshakeStatus",
            "locals": {
              "self": "<class 'twisted.protocols.tls.BufferingTLSTransport'>"
            }
          },
          {
            "code": "if _representsEOF(reason.value):",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py",
            "argspec": [
              "self",
              "reason"
            ],
            "lineno": 377,
            "method": "_tlsShutdownFinished",
            "locals": {
              "self": "<class 'twisted.protocols.tls.BufferingTLSTransport'>",
              "reason": "<class 'twisted.python.failure.Failure'>"
            }
          },
          {
            "code": "errorQueue = exceptionObject.args[0]",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py",
            "argspec": [
              "exceptionObject"
            ],
            "lineno": 121,
            "method": "_representsEOF",
            "locals": {
              "exceptionObject": "<class 'OpenSSL.SSL.ZeroReturnError'>"
            }
          }
        ],
        "exception": {
          "message": "tuple index out of range",
          "class": "IndexError"
        }
      },
      {
        "frames": [
          {
            "code": "self._tlsConnection.do_handshake()",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/twisted/protocols/tls.py",
            "argspec": [
              "self"
            ],
            "lineno": 247,
            "method": "_checkHandshakeStatus",
            "locals": {
              "self": "<class 'twisted.protocols.tls.BufferingTLSTransport'>"
            }
          },
          {
            "code": "self._raise_ssl_error(self._ssl, result)",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/OpenSSL/SSL.py",
            "argspec": [
              "self"
            ],
            "lineno": 2193,
            "method": "do_handshake",
            "locals": {
              "self": "<class 'OpenSSL.SSL.Connection'>",
              "result": -1
            }
          },
          {
            "code": "raise ZeroReturnError()",
            "filename": "/usr/src/app/.venv/lib/python3.11/site-packages/OpenSSL/SSL.py",
            "argspec": [
              "self",
              "ssl",
              "result"
            ],
            "lineno": 1805,
            "method": "_raise_ssl_error",
            "locals": {
              "ssl": "<class '_cffi_backend.__CDataGCP'>",
              "self": "<class 'OpenSSL.SSL.Connection'>",
              "result": -1,
              "error": 6
            }
          }
        ],
        "exception": {
          "message": "",
          "class": "ZeroReturnError"
        }
      }
    ]
  }
```

**Comment by glyph** [comment]
> I can provide a detailed stack-trace. Does this help clarify things?

Thanks, I don't *think* this is quite enough info, but it is at least a place to get started :)

**Comment by OliverFarren** [comment]
> > I can provide a detailed stack-trace. Does this help clarify things?
> 
> Thanks, I don't _think_ this is quite enough info, but it is at least a place to get started :)

I assume i'm missing something, but isn't the issue that the `_representsEOF` method assumes that the incoming exception object was instantiated with args but that isn't always the case?

**Comment by glyph** [comment]
> I assume i'm missing something, but isn't the issue that the `_representsEOF` method assumes that the incoming exception object was instantiated with args but that isn't always the case?

It is, but the question is *why* isn't that the case. I absolutely recognize the tendency here to say "oh, well, openssl is weird sometimes, let's just add more code to handle this bizarre new condition it threw at us" but this *is* a security-critical code path, and we should understand *what* exactly OpenSSL is trying to say (beyond "we are bad at software") by giving us an exception in this shape.

**Comment by OliverFarren** [comment]
I see, so the expectation is that an OpenSSL.SSL.ZeroReturnError would get invoked with arguments and this seems like a divergence from the norm? 

On our side, this gets caught by our Error tracking system and occurs relatively rarely, perhaps 10 times over 14 days of our scrapy based application running in our production environment. Each time we get a stack trace similar to the one above. At the application level i'm not sure how I could reproduce. I can try running our service for a few days in a debug environment with some breakpoints set, that's not guaranteed to reproduce the set of conditions that reach this but i'd be happy to try. I'd need some guidance on where to set some breakpoints to inspect as i'm not too familiar with the inner workings of twisted or OpenSSL. 

In anycase, let me know if there's anything I can do to help in understanding the issue more

**Comment by OliverFarren** [comment]
Looking at the [current OpenSSL](https://github.com/pyca/pyopenssl/blob/d9f2c46de70c1aee20a4309424d9f506b7aae68e/src/OpenSSL/SSL.py#L1805) it seems like the ZeroReturnError is always raised without args?

**Comment by OliverFarren** [comment]
Perhaps a change to version of OpenSSL and/or cryptography has changed the flow? 

Looking at our version history, in the last 6 months we've gone from using
pyopenssl == 23.3.0 to 24.1.0
[changelog](https://www.pyopenssl.org/en/latest/changelog.html)
And
cryptography == 41.0.5 to 42.0.7
[changelog](https://cryptography.io/en/latest/changelog/)
And
twisted == 22.10.0 to 23.10.0

Our dependency on using 23.10.0 and not the newest is all down to scrapy's constraints.

**Comment by glyph** [comment]
> Looking at the [current OpenSSL](https://github.com/pyca/pyopenssl/blob/d9f2c46de70c1aee20a4309424d9f506b7aae68e/src/OpenSSL/SSL.py#L1805) it seems like the ZeroReturnError is always raised without args?

Yes, it always has been, and the one place in the code where Twisted *explicitly* handles a `ZeroReturnError` (i.e. here: https://github.com/twisted/twisted/blob/c465c460c68bec3cf481c2815d1039da6796eb03/src/twisted/protocols/tls.py#L289-L295) it doesn't expect it to have arguments. The question is, what operation is raising a `ZeroReturnError` elsewhere, and why is it doing that unexpectedly?  Is the issue that we get it from `do_handshake` sometimes, rather than `recv`? I don't see the line numbers lining up here, but it seems likely that's what's going on.  What does a `ZeroReturnError` from `do_handshake`, specifically, mean?

**Comment by OliverFarren** [comment]
Not hot on this stuff but could there have been a change to OpenSSL? Seems possible looking at some PR's

- https://github.com/openssl/openssl/pull/21780
- https://github.com/openssl/openssl/pull/21618
- https://github.com/search?q=repo%3Aopenssl%2Fopenssl+SSL_ERROR_ZERO_RETURN&type=pullrequests

**Comment by OliverFarren** [comment]
Perhaps also relevant in https://www.openssl.org/docs/man3.0/man7/migration_guide.html under TLS changes

> New SSL option SSL_OP_IGNORE_UNEXPECTED_EOF
>
> The SSL option SSL_OP_IGNORE_UNEXPECTED_EOF is introduced. If that option is set, an unexpected EOF is ignored, it pretends a close notify was received instead and so the returned error becomes SSL_ERROR_ZERO_RETURN.

**Comment by adiroiban** [comment]
I think that pyOpenSSL is fine.

`ZeroReturnError` is a very specific exception and has no extra text/details/message.... similar to WantRead or WantWrite.

I think that the issue is in Twisted code, as Twisted assumes that any error has an argument.

https://github.com/twisted/twisted/blob/c465c460c68bec3cf481c2815d1039da6796eb03/src/twisted/protocols/tls.py#L113-L123

Just like we have special handling for `SysCallError`, we should also have special handling for other errors like `ZeroReturnError`

-------

More info about the conditions for which this exception are raised can be found at https://www.openssl.org/docs/man3.3/man3/SSL_get_error.html

> SSL_ERROR_ZERO_RETURN
> The TLS/SSL peer has closed the connection for writing by sending the close_notify alert. No more data can be read. Note that SSL_ERROR_ZERO_RETURN does not necessarily indicate that the underlying transport has been closed.
>
> This error can also appear when the option SSL_OP_IGNORE_UNEXPECTED_EOF is set. See [SSL_CTX_set_options(3)](https://www.openssl.org/docs/man3.3/man3/SSL_CTX_set_options.html) for more details.


--------

I don't know how easy is to replicate this error in an unit test

**Comment by OliverFarren** [comment]
> I think that pyOpenSSL is fine.
> 
> `ZeroReturnError` is a very specific exception and has no extra text/details/message.... similar to WantRead or WantWrite.
> 
> I think that the issue is in Twisted code, assuming that any error has an argument
> 
> https://github.com/twisted/twisted/blob/c465c460c68bec3cf481c2815d1039da6796eb03/src/twisted/protocols/tls.py#L113-L123
> 
> Just like we have special handling for `SysCallError` we should also have special handling for other errors like `ZeroReturnError`
> 
> More info about the conditions for which this exception are raised can be found at https://www.openssl.org/docs/man3.3/man3/SSL_get_error.html
> 
> > SSL_ERROR_ZERO_RETURN
> > The TLS/SSL peer has closed the connection for writing by sending the close_notify alert. No more data can be read. Note that SSL_ERROR_ZERO_RETURN does not necessarily indicate that the underlying transport has been closed.
> > This error can also appear when the option SSL_OP_IGNORE_UNEXPECTED_EOF is set. See [SSL_CTX_set_options(3)](https://www.openssl.org/docs/man3.3/man3/SSL_CTX_set_options.html) for more details.
> 
> I don't know how easy is to replicate this error in an unit test

I agree, I think it was more around understanding what's changed.

**Comment by OliverFarren** [comment]
@glyph, any more thoughts on this?

**Comment by glyph** [comment]
> @glyph, any more thoughts on this?

It sounds like we have a handle on what's going on here, at least vaguely. I'd like to see a test for this; it *should* be possible to introduce an EOF by using a MemoryReactor style test. But if that's too much effort or we can't make it work where we think we should, a fake `do_handshake` wrapper that just directly tests we handle this case is fine.

**Comment by jcoetzee** [comment]
Has there been any progress on this issue?

**Comment by glyph** [comment]
@jcoetzee Nope. If there is you'll see it here. Do you have any interest in writing a test for it?
